polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1211,31 +1252,64 @@ impl<T: Config> Pallet<T> {
 		ensure!(assets.len() <= MAX_ASSETS_FOR_TRANSFER, Error::<T>::TooManyAssets);
 		let value = (origin_location, assets.into_inner());
 		ensure!(T::XcmReserveTransferFilter::contains(&value), Error::<T>::Filtered);
-		let (origin_location, assets) = value;
-		let context = T::UniversalLocation::get();
-		let fees = assets
-			.get(fee_asset_item as usize)
-			.ok_or(Error::<T>::Empty)?
-			.clone()
-			.reanchored(&dest, context)
-			.map_err(|_| Error::<T>::CannotReanchor)?;
-		let max_assets = assets.len() as u32;
-		let assets: MultiAssets = assets.into();
-		let xcm = Xcm(vec![
-			BuyExecution { fees, weight_limit },
-			DepositAsset { assets: Wild(AllCounted(max_assets)), beneficiary },
-		]);
-		let mut message = Xcm(vec![
-			SetFeesMode { jit_withdraw: true },
-			TransferReserveAsset { assets, dest, xcm },
-		]);
-		let weight =
-			T::Weigher::weight(&mut message).map_err(|()| Error::<T>::UnweighableMessage)?;
-		let hash = message.using_encoded(sp_io::hashing::blake2_256);
-		let outcome =
-			T::XcmExecutor::execute_xcm_in_credit(origin_location, message, hash, weight, weight);
-		Self::deposit_event(Event::Attempted { outcome });
-		Ok(())
+		let (origin_location, mut assets) = value;
+
+		if fee_asset_item as usize >= assets.len() {
+			return Err(Error::<T>::Empty.into())
+		}
+		let fees = assets.swap_remove(fee_asset_item as usize);
+		let fees_transfer_type =
+			<T::XcmExecutor as AssetTransferSupport>::determine_for(&fees, &dest)
```

Comment:
```suggestion
			T::XcmExecutor::determine_for(&fees, &dest)
```

it works without `<T::XcmExecutor as AssetTransferSupport>`, I dont know if we want to be explicit here